Create import_region method for loading region objects

javerbukh opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request creates an `import_region` method that loads a region object with `combination_mode`.
Some todo:
- [x] Implement loading region file using https://github.com/spacetelescope/jdaviz/blob/dfb4bfa9cf41a2129981c5c2f11ee5d4dca8677f/jdaviz/core/helpers.py#L651
- [x] `self.combination_mode` should use the `Select` (`SubsetSelect`?) class to be able to take advantage of the `choices` and `selected` attributes
- [x] Connect `self.combination_mode` to the subset mode selection in the toolbar
- [x] Add tests
- [x] Add documentation for new API

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #<Issue Number>

### Change log entry

- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,7 +677,158 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
 
-    def _import_spectral_regions(self, spec_region, mode=NewMode, viewer=None):
+    def import_region(self, region, **kwargs):
+        return_bad_regions = kwargs.pop('return_bad_regions', None)
```
Why all the popping? Why not define these keywords explicitly instead of a blind `**kwargs`? Given that this is a public API, it is important for users to know all their options in the docstring as well (when you write the docstring that Kyle asked for).

pllim commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,7 +677,158 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+                region_format = kwargs.pop('region_format', None)
+                try:
+                    raw_regs = Regions.read(region, format=region_format)
+                except:  # noqa
```
```suggestion
                except Exception:  # nosec
```

pllim commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,7 +677,158 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+        if (isinstance(regions, SpectralRegion) or
+                (isinstance(regions, list) and isinstance(regions[0], SpectralRegion))):
+            mode = kwargs.pop('mode', None)
+            self._import_spectral_regions(regions, mode)
+            return
```
1. Given that you `return` here, all the spatial region imports (if you still want them locally scoped) should happen after this.
2. You should `return bad_regions` here to be consistent with spatial loader behavior, which means your `_import_spectral_regions` should give you back `bad_regions` like the spatial one does.

```suggestion
            return self._import_spectral_regions(regions, mode)
```

pllim commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,7 +677,158 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+            snack_color = "warning"
+        else:
+            snack_color = "success"
+        self.app.hub.broadcast(SnackbarMessage(
+            f"Loaded {n_loaded}/{n_reg_in} regions, max_num_regions={max_num_regions}, "
+            f"bad={n_reg_bad}", color=snack_color, timeout=8000, sender=self.app))
+        if return_bad_regions:
+            return bad_regions
```
This behavior should be emulated by the spectral region loader too, to be consistent.

pllim commented on jdaviz/configs/default/plugins/subset_plugin/tests/test_subset_plugin.py:
```diff
@@ -177,23 +176,44 @@ def test_circle_recenter_linking(roi_class, subset_info, imviz_helper, image_2d_
+    cubeviz_helper.load_data(spectrum1d_cube)
+    plg = cubeviz_helper.plugins['Subset Tools']._obj
+    s = SpectralRegion(5*u.um, 6*u.um)
+    s.write()
```
Please write to `tmp_path`. Otherwise, this test will pollute the source dir when we run the tests locally.

pllim commented on jdaviz/configs/imviz/plugins/parsers.py:
```diff
@@ -123,7 +123,8 @@ def parse_data(app, file_obj, ext=None, data_label=None,
         elif file_obj_lower.endswith('.reg'):
             # This will load DS9 regions as Subset but only if there is already data.
-            app._jdaviz_helper.load_regions_from_file(file_obj)
+            # app.plugins['Subset Tools'].import_region(file_obj)
```
Is there a reason to keep the commented line?

```suggestion
```

pllim commented on jdaviz/configs/imviz/tests/test_regions.py:
```diff
@@ -168,71 +168,73 @@ def test_regions_annulus_from_load_data(self):
             'data/ds9.fits.reg', package='regions.io.ds9.tests')
         self.arr = np.ones((1024, 1024))
         self.raw_regions = Regions.read(self.region_file, format='ds9')
+        self.imviz = imviz_helper
```
This changes the fundamental behavior of the test class. I wanted each method to have a new Imviz instance, but this change makes them all share the same instance. So hardcoded check like "Subset 1" is no longer reliable. Please revert this change.

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -662,6 +663,7 @@ def default_viewer(self):
         (e.g., "imviz-0" or "cubeviz-0")."""
         return self._default_viewer.user_api
 
+    @deprecated(since="4.0", alternative="subset_plugin.import_region")
```
```suggestion
    @deprecated(since="4.0", alternative="viz.plugins['Subset Tools'].import_region")
```

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -694,6 +696,7 @@ def load_regions_from_file(self, region_file, region_format='ds9', max_num_regio
         raw_regs = Regions.read(region_file, format=region_format)
         return self.load_regions(raw_regs, max_num_regions=max_num_regions, **kwargs)
 
+    @deprecated(since="4.0", alternative="subset_plugin.import_region")
```
```suggestion
    @deprecated(since="4.0", alternative="viz.plugins['Subset Tools'].import_region")
```

pllim commented on notebooks/CubevizExample.ipynb:
Please revert your irrelevant changes here. Looks like you forgot to clear the notebook output before checking things in.

pllim commented on notebooks/SpecvizExample.ipynb:
Please revert your irrelevant changes here. Looks like you forgot to clear the notebook output before checking things in.

pllim reviewed (changes requested):
All the existing tests using `apply_roi` should be updated to use this new API too?

Is there any benefit to use locally scoped imports vs importing everything you need at the top of the module?

Other comments as follow.

pllim commented:
Also:

1. Add a change log
2. Fix error in `jdaviz/configs/cubeviz/plugins/tests/test_tools.py`: `AssertionError: assert 3 == 4`

Thanks!

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -676,10 +850,30 @@ def _import_spectral_regions(self, spec_region, mode=NewMode, viewer=None):
+    def _combination_selected_updated(self, change):
+        self.app.session.edit_subset_mode.mode = SUBSET_MODES[self.combination_mode.selected]
+
+    def _update_combination_mode(self):
```
this is only ever called when linking is changed, but needs to be two-way synced, so that changing the mode in the UI (the glue dropdown) or via internal glue API, is reflected in `self.combination_mode`

kecnry replied:
has this been addressed?  When changing the combination mode from the UI, I do not see the internal traitlet being updated to match (although changing the `combination_mode.selected` does reflect in the UI via `_combination_selected_updated`).

javerbukh commented:
@pllim @kecnry I believe I have addressed all review comments except the ones about `_import_spectral_regions` behaving exactly like the spatial subset creation. I need to put this PR down until the 26th so please feel free to handle that additional work however you see fit.

pllim commented:
> I need to put this PR down until the 26th

And this blocks 4.0?

kecnry commented:
> And this blocks 4.0?

This is not listed as required for 4.0.  In fact, if we _do_ merge it, we may want to consider postponing exposing the API so that we continue to iterate on the public API during the follow-up efforts.

bmorris3 reviewed (commented):
This looks good. Should we also change the example notebooks (ImvizExample, CubevizExample, ImvizDitheredExample, imviz_advanced_aper_phot, imviz_color_mpl, imviz_edit_subset_programmatic, cubeviz-cropped) which still have calls to `helper.load_regions`?

javerbukh commented:
So add a `4.1 (unreleased)` to the `CHANGES.rst` file and a new 4.1 milestone?

pllim commented:
The change log might have to wait, otherwise you will run into conflict. I can remilestone in the meantime.

pllim commented:
I still find 144 instances of `apply_roi` that can be replaces with your new API. Do you have plans to address that in this PR? The point of this API is so user can stop using `apply_roi`, so I think our own code should be consistent internally.

```
$ grep apply_roi -rl jdaviz/
jdaviz/core/helpers.py
jdaviz/core/tools.py
jdaviz/core/tests/test_helpers.py
jdaviz/core/tests/test_data_menu.py
jdaviz/core/tests/test_template_mixin.py
jdaviz/tests/test_subsets.py
jdaviz/configs/cubeviz/plugins/spectral_extraction/tests/test_spectral_extraction.py
jdaviz/configs/cubeviz/plugins/tests/test_regions.py
jdaviz/configs/cubeviz/plugins/tests/test_parsers.py
jdaviz/configs/cubeviz/plugins/tests/test_cubeviz_helper.py
jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py
jdaviz/configs/specviz/tests/test_helper.py
jdaviz/configs/specviz/plugins/line_analysis/tests/test_line_analysis.py
jdaviz/configs/default/plugins/data_quality/tests/test_data_quality.py
jdaviz/configs/default/plugins/model_fitting/tests/test_plugin.py
jdaviz/configs/default/plugins/export/tests/test_export.py
jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py
jdaviz/configs/default/plugins/subset_plugin/tests/test_subset_plugin.py
```

javerbukh commented:
This PR is not about replacing `apply_roi`, it is about creating an API to create subsets and import them from a file using the subset plugin. Please correct me if I'm wrong @kecnry @camipacifici .

kecnry commented:
I personally think its ok if the base logic lives in an internal call.  The plugin provides a UI and API interface to the user, but that doesn't always make sense for all of our internal uses (and likely would add overhead).  That said, we probably should avoid duplicated logic whenever possible (I haven't looked at the diff to see if this is copying things from `apply_roi` or just adding a user-friendly wrapper around it.

If we want a consistent API for both internal calls and users, then we can mirror any changes to the calling signature of `apply_roi` or even call it something else, but I would be hesitant to make calls to a plugin from the app whenever possible.

kecnry commented:
If we decide to replace `apply_roi` entirely, I think that could be a follow-up ticket, the scope of this has already extended quite a bit beyond the original scoped API.

pllim commented:
(Shouldn't we be using the API that we want users to use in test code?)

javerbukh commented:
I can't replace all calls to `apply_roi` but I can update the calls that are in tests.

javerbukh commented:
@pllim @kecnry Tests have been updated and fixed, please review at your earliest convenience.

javerbukh commented:
@kecnry I realized that many of the tests assume `viz.plugins['Subset Tools'].import_region` is public API so I reverted the commit. We can wait to merge this in that case.

pllim commented:
Do we just want to branch v4.0.x now and then merge this into `main`?

pllim commented:
My ticket too longer than expected. I will review Tuesday. Sorry for the delay.

pllim commented:
Note to self: This shall not be merged until v4.0.x is branched out. Then Jesse is going to add change log to v4.1 section.

pllim commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +699,348 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+        from jdaviz.core.region_translators import regions2roi, aperture2regions
+
+        # If user passes in one region obj instead of list, try to be smart.
+        if (isinstance(regions, SpectralRegion) or
+                (isinstance(regions, list) and isinstance(regions[0], SpectralRegion))):
+            mode = kwargs.pop('mode', None)
+            return self._import_spectral_regions(regions, mode, max_num_regions, return_bad_regions)
```
I remember @kecnry was concerned that `apply_roi` is no longer used within the public API to load spectral region, causing deviation between external and internal usage. Indeed, I do not see `_import_spectral_regions` use `apply_roi` at all because `_import_spectral_regions` is only called from image regions. So this concern still stands?

In fact, `create_new_after` is also not used?

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -671,6 +672,7 @@ def default_viewer(self):
         (e.g., "imviz-0" or "cubeviz-0")."""
         return self._default_viewer.user_api
 
+    @deprecated(since="4.0", alternative="viz.plugins['Subset Tools'].import_region")
```
When a change log is finally added, it needs to clarify that `max_num_regions=20` is no longer automatically set if loading image regions from file.

pllim commented on notebooks/concepts/imviz_edit_subset_programmatic.ipynb:
```diff
@@ -68,31 +68,31 @@
    "outputs": [],
    "source": [
-    "subset_tool = imviz.plugins['Subset Tools']"
+    "imviz.default_viewer.zoom_level = 5"
```
Where did the zoom level come from?

pllim reviewed (commented):
1. Should `import_region` take ROI as input at all? ROI object is internal while the API is public. Or is this translation oversight? I think the ROI code should be converted to a proper circular region in the affected tests?
2. If the tests are meant to test the public API, should `subset_plugin = cubeviz_helper.plugins['Subset Tools']` be used instead of `subset_plugin = cubeviz_helper.plugins['Subset Tools']._obj` in the test code?
3. In the test code where `import_region(SpectralRegion(...))` is blindly using unit set in Unit Conversion plugin, I feel like the unit should be hardcoded instead. When this API is used, usually the user already have a unit in mind and won't refer to the Unit Conversion plugin at all. So reading the unit off Unit Conversion plugin does not reflect a realistic use case.

camipacifici commented:
Hi @javerbukh, nice work!
I took this for a spin and have a couple of comments:
- For me `create_new_after=True` did nothing. Calling `import_region` would still overwrite the same subset with or without it. Maybe I am missing something.
- I tried the workflow where I create a list of regions and load all of them at once with
`regionlist = SpectralRegion([(1.5*u.um, 2.5*u.um),(3.5*u.um, 4.5*u.um)])`
`subtools.import_region(regionlist)`
Then I remembered I needed `subtools.combination_mode = 'new'` to be able to do this, which is fine, but I would argue that the default should be that after creating a subset, it creates a new one instead of overwriting the existing one.
What do you think?

javerbukh commented:
@camipacifici For your second use case, doing `subtools.import_region(regionlist, create_new_after=True)` would do what you are expecting.

camipacifici commented:
`create_new_after` does nothing on my side. I do not understand why. But I second Kyle's comment. It feels more natural to act on the current operation than on the next.

javerbukh commented:
@camipacifici `create_new_after` has been renamed to `create_as_new` and now behaves as you would expect, including for spectral subsets. Please let me know if that works for you!

camipacifici commented:
Thank you for the new updates!
I leave here something for discussion. My workflow is in specviz where I want to create three subsets
`region = SpectralRegion(1*u.um, 2*u.um) #alone`
`region2 = SpectralRegion(3*u.um, 4*u.um) #alone`
`regionlist = SpectralRegion([(1.5*u.um, 2.5*u.um),(3.5*u.um, 4.5*u.um)]) #2 wavelength ranges in same subset`
This does the job:
`subtools._obj.import_region(region, combination_mode='new')`
`subtools._obj.import_region(region2, combination_mode='new')`
`subtools._obj.import_region(regionlist, combination_mode=['new','or'])`
I wonder if `combination_mode='new'` should be the default for the first subset, to be able to achieve the same thing with this instead
`subtools._obj.import_region(region)`
`subtools._obj.import_region(region2)`
`subtools._obj.import_region(regionlist, combination_mode='or')`
What do people think?

Setting `subtools._obj.combination_mode = 'new'` at the beginning gives me a long traceback at the first `import_region`, but maybe I am doing something wrong there.

javerbukh commented:
The first `import_region` call should create a new subset by default and all following calls will need the `combination_mode='new'` parameter. The reason for that is our `viewer.apply_roi` API behaves the same as the GUI where every call after the first just updates the selected subset. Creating a new subset with each call is what `helper.load_regions` does. This method can do both but the default glue and `apply_roi` implementation is also the default here.

javerbukh commented:
Sorry, forgot to address this point! You need to do `subtools._obj.combination_mode.selected = 'new'` because `combination_mode` is a `SelectPluginComponent` object. You can also do `combination_mode.choices` to see options for what to select.

kecnry commented:
ah ok, good to know its only that!.  (But note that once exposed to the user API, the `.selected` will no longer be required)

pllim reviewed (approved):
I think we're at the point where this has dragged on long enough that it should just go in so we can move on to the follow-up tickets. CI is green, so approving. Thanks for your patience and persistence!

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -34,10 +41,31 @@
+    * :meth:`~jdaviz.core.template_mixin.PluginTemplateMixin.show`
+    * :meth:`~jdaviz.core.template_mixin.PluginTemplateMixin.open_in_tray`
+    * :meth:`~jdaviz.core.template_mixin.PluginTemplateMixin.close_in_tray`
+    * ``combination_mode`` (:class:`~jdaviz.core.template_mixin.SelectPluginComponent`):
+      The method used for adding or combining subsets.
+    * :meth:`import_region`
```
These aren't yet exposed in the user API

```suggestion
```

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+        self.app.session.edit_subset_mode.mode = SUBSET_MODES_PRETTY[change['new']]
+
+    def _update_combination_mode(self):
+        for key, value in SUBSET_MODES_PRETTY.items():
+            if self.app.session.edit_subset_mode.mode == value:
+                self.combination_mode.selected = key
+                return
```
if this would ever need to be done in more than one place, we could have an inverted dictionary defined at the top (although then the dictionaries should probably be named something that makes it clear which is which direction).

```
SUBSET_TO_PRETTY = {v: k for k,v in SUBSET_MODES_PRETTY.items()}
```

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+        if len(self.app.data_collection) == 0:
+            raise ValueError('Cannot load regions without data.')
+
+        from photutils.aperture import (CircularAperture, SkyCircularAperture,
+                                        EllipticalAperture, SkyEllipticalAperture,
+                                        RectangularAperture, SkyRectangularAperture,
+                                        CircularAnnulus, SkyCircularAnnulus)
+        from regions import (Regions, CirclePixelRegion, CircleSkyRegion,
+                             EllipsePixelRegion, EllipseSkyRegion,
+                             RectanglePixelRegion, RectangleSkyRegion,
+                             CircleAnnulusPixelRegion, CircleAnnulusSkyRegion)
+        from jdaviz.core.region_translators import regions2roi, aperture2regions
```
is there a reason these are not imported at the top-level?

javerbukh replied:
This is from the original code so probably a question for @pllim .

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+                             CircleAnnulusPixelRegion, CircleAnnulusSkyRegion)
+        from jdaviz.core.region_translators import regions2roi, aperture2regions
+
+        if not isinstance(regions, (list, tuple, Regions, SpectralRegion)):
```
why is SpectralRegion in this list?  Does it automatically handle iterating over itself?

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+        else:
+            data = self.app.data_collection[refdata_label]
+
+        # TODO: Make this work for data cube.
```
is this a new TODO comment or copied from the previous region import code?

javerbukh replied:
Copied from previous, I'll remove it.

kecnry replied:
does it work for cubes?  If not, let's make sure we have a ticket (but safe to ignore here if this was from moved code)

javerbukh replied:
Seems like this works, although the solution is hacky (per https://github.com/glue-viz/glue-astronomy/issues/75 ), maybe @pllim can clarify on the status. Subset creation works with cubes so I think this is safe to remove but I'll wait for Pey-Lian's answer.

pllim replied:
@rosteen reworked some spectrum WCS stuff recently so I'll defer to him. Or was it @dhomeier ? As long as the tests don't fail and results don't change, whatever works is fine by me.

rosteen replied:
If it works it works 🤷 I don't recall changing the spectrum WCS stuff recently in a way that would affect this.

dhomeier replied:
That issue has been closed by https://github.com/glue-viz/glue-astronomy/issues/75#issuecomment-2359390639, and the translator would only use a `PaddedSpectrumWCS` if the cube had a 1D WCS, which I assume is not the case here.

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+                self.app.session.edit_subset_mode.edit_subset = None  # No overwrite next iteration
+                self.app.session.edit_subset_mode.mode = SUBSET_MODES_PRETTY['new']
+            elif combo_mode:
+                self.combination_mode.selected = combo_mode
```
so if a list of regions is passed with a list of combination modes but an entry is invalid, the previous entries will be loaded and then an error will be raised here that it is not valid?  Can/should we validate the entire list in advance (possibly as follow-up if necessary)?

javerbukh replied:
Validate the list of combination modes? Yes that should be a quick addition.

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +697,299 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+            A string which represents a ``regions`` or ``SpectralRegion`` file.
+            If given as a list, it can only contain spectral or non-spectral regions, not both.
+
+        combination_mode : list, str, or `None`
+            The way that regions are created or combined. If a list, then it must be the
+            same length as regions. If `None`, then it will follow the default glue
+            functionality for subset creation.
```
could/should we restore the original combination mode after this either completes or raises an exception?

javerbukh replied:
My thought would be the combination mode stays the last entry in the list but I am interested to hear what would make the most sense for a user.

kecnry replied:
ok, let's discuss and decide with the team.  Whatever the decision, its probably worth explaining here in the API docs so the users know what to expect.

javerbukh commented:
@kecnry Waiting to hear from @rosteen on whether the line about checking a cube's wcs is still valid but otherwise I addressed the review comments. I tested changing the edit mode using the UI and API and it seemed to update correctly. Please let me know if you have a workflow I can use to test that behavior.

As for the comment on combination_mode reverting to default after `import_region` is called, I think we should defer that discussion to a follow-up. I created ticket for that in the subset plugin epic on Jira.

kecnry commented:
After digging into this more, this might just be a limitation of how glue handles the mode dropdown vs the actual mode that will be used (when there is no active subset selected, there will be a new created subset regardless of the mode dropdown).   If/when we want to expose a UI element in the plugin, we'll definitely need to think about this a little more.  I'm not sure if/how we should deal with this for the exposed API though 🤔

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -34,10 +50,30 @@
+    'andnot': AndNotMode,
+}
+SUBSET_TO_PRETTY = {v: k for k, v in SUBSET_MODES_PRETTY.items()}
+COMBO_OPTIONS = ['new', 'replace', 'or', 'and', 'xor', 'andnot']
```
```suggestion
COMBO_OPTIONS = list(SUBSET_MODES_PRETTY.keys())
```

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -93,6 +132,17 @@ def __init__(self, *args, **kwargs):
         align_by = getattr(self.app, '_align_by', None)
         self.display_sky_coordinates = (align_by == 'wcs' and not self.multiselect)
 
+        combination_options = list(SUBSET_MODES_PRETTY)
+        self.combination_mode = SelectPluginComponent(self,
+                                                      items='combination_items',
+                                                      selected='combination_selected',
+                                                      manual_options=combination_options)
```
```suggestion
        self.combination_mode = SelectPluginComponent(self,
                                                      items='combination_items',
                                                      selected='combination_selected',
                                                      manual_options=combo_options)
```

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +705,290 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+            Default is to load everything.  If you are providing a large file/list
+            input for ``region``, it is recommended
+
+        refdata_label : str or `None`
```
I wonder if we could just do `data_label` - but we can revise this anytime before making public, so can ignore for now.

kecnry commented on jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py:
```diff
@@ -654,32 +705,290 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
+            if os.path.exists(region):
+                from regions import Regions
+                region_format = kwargs.pop('region_format', None)
+                try:
+                    raw_regs = Regions.read(region, format=region_format)
+                except Exception:  # nosec
+                    raw_regs = SpectralRegion.read(region)
```
This is fine for now, but I wonder if there's any way we could predict in advance instead of try/except.  As it is now, checking the extension would probably work, but in the future we may have the same extension supported between spectral + spatial, so that isn't a longterm solution.

kecnry reviewed (approved):
LGTM now - any remaining items can be iterated as we build on top of this before we make the API exposed as public.  Thanks for working through all these iterations!

If we want this merged before @javerbukh gets back, I think we could just accept the few small code suggestions I listed and then merge.

pllim commented:
> we could just accept the few small code suggestions I listed and then merge.

+1
